db/version_set.cc

```diff
@@ -1317,6 +1698,163 @@ void Version::Get(const ReadOptions& read_options, const LookupKey& k,
   }
 }
 
+void Version::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
+                       ReadCallback* callback, bool* is_blob) {
+  PinnedIteratorsManager pinned_iters_mgr;
+
+  // Pin blocks that we read to hold merge operands
+  if (merge_operator_) {
+    pinned_iters_mgr.StartPinning();
+  }
+
+  // Even though we know the batch size won't be > MAX_KEYS_ON_STACK,
+  // use autovector in order to avoid unnecessary construction of GetContext
+  // objects, which is expensive
+  autovector<GetContext, MultiGetContext::MAX_KEYS_ON_STACK> get_ctx;
+  for (auto iter = range->begin(); iter != range->end(); ++iter) {
+    assert(iter->s->ok() || iter->s->IsMergeInProgress());
+    get_ctx.emplace_back(
+        user_comparator(), merge_operator_, info_log_, db_statistics_,
+        iter->s->ok() ? GetContext::kNotFound : GetContext::kMerge, iter->ukey,
+        iter->value, nullptr, &(iter->merge_context),
+        &iter->max_covering_tombstone_seq, this->env_, &iter->seq,
+        merge_operator_ ? &pinned_iters_mgr : nullptr, callback, is_blob);
+    iter->get_context = &get_ctx.back();
+  }
+
+  MultiGetRange file_picker_range(*range, range->begin(), range->end());
+  FilePickerMultiGet fp(
+      storage_info_.files_, &file_picker_range,
+      &storage_info_.level_files_brief_, storage_info_.num_non_empty_levels_,
+      &storage_info_.file_indexer_, user_comparator(), internal_comparator());
+  FdWithKeyRange* f = fp.GetNextFile();
+
+  while (f != nullptr) {
+    MultiGetRange file_range = fp.CurrentFileRange();
```

Comment:
Is making a copy of `MultiGetRange` intended?

Reply:
Good catch! A reference will do here